src/hotspot/share/prims/jvmtiEnvBase.cpp

```diff
@@ -1529,6 +1529,64 @@ JvmtiEnvBase::is_in_thread_list(jint count, const jthread* list, oop jt_oop) {
   return false;
 }
 
+class VM_InitNotifyJvmtiEventsMode : public VM_Operation {
+private:
+  bool _enable;
+
+  int count_transitions() {
+    int count = 0;
+    if (_enable) {
+      for (JavaThread* jt : ThreadsListHandle()) {
+        if (jt->is_in_VTMS_transition()) {
+          count++;
+        }
+      }
+    }
+    return count;
+  }
+
+public:
+  VMOp_Type type() const { return VMOp_InitNotifyJvmtiEventsMode; }
+  bool allow_nested_vm_operations() const { return false; }
+  VM_InitNotifyJvmtiEventsMode(bool enable) : _enable(enable) {}
+
+  void doit() {
+    int trans_count = count_transitions();
+
+    JvmtiVTMSTransitionDisabler::set_VTMS_notify_jvmti_events(_enable);
+    JvmtiVTMSTransitionDisabler::set_VTMS_transition_count(trans_count);
+  }
+};
+
+// must be called in thread-in-native mode
+bool
+JvmtiEnvBase::enable_virtual_threads_notify_jvmti() {
+  if (!Continuations::enabled()) {
+    return false;
+  }
+  if (JvmtiVTMSTransitionDisabler::VTMS_notify_jvmti_events()) {
+    return false; // already enabled
+  }
+  VM_InitNotifyJvmtiEventsMode op(true);
+  VMThread::execute(&op);
+  return true;
+}
+
+// must be called in thread-in-native mode
+bool
+JvmtiEnvBase::disable_virtual_threads_notify_jvmti() {
+  if (!Continuations::enabled()) {
+    return false;
+  }
+  if (JvmtiVTMSTransitionDisabler::VTMS_notify_jvmti_events()) {
```

Comment:
shouldn't be
 if (!JvmtiVTMSTransitionDisabler::VTMS_notify_jvmti_events()) {
here?

Reply:
Fixed it now and discovered some issues which have not seen before.
It occurred that disabling notifyJvmti events when virtual threads are executed is unsafe.
It is not easy to fix these issue and there is no real need to.
So that I've refactored the test to do multiple testing cycles.
Each testing cycle disables notifyJvmti events when there are no virtual threads executed then starts a number of threads, then enables notifyJvmti events and shuts down the virtual threads.
The test also extended to post more JVMTI events:  `VirtualThreadStart`, `VirtualThreadEnd`, `ThreadStart` and `ThreadEnd`.
Also, I saw some intermittent crashes with double-deallocation of JvmtiThreadState's which belongs to vthreads.
So that, I have extend the VM_InitNotifyJvmtiEventsMode to do more corrections.
Will push my fixes after my mach5 test runs are finished.